rust/src/soe/mod.rs

```diff
@@ -1,46 +1,50 @@
-use crate::FifoWindow;
-use alga::general::AbstractGroup;
-use alga::general::Operator;
 use std::collections::VecDeque;
-use std::marker::PhantomData;
+
+use alga::general::AbstractMagma;
+use alga::general::Identity;
+use alga::general::TwoSidedInverse;
+
+use crate::FifoWindow;
+use crate::ops::AggregateOperator;
+use crate::ops::AggregateGroup;
 
 #[derive(Clone)]
-pub struct SoE<Value, BinOp>
+pub struct SoE<BinOp>
 where
-    Value: AbstractGroup<BinOp> + Clone,
-    BinOp: Operator,
+    BinOp: AggregateGroup<BinOp> + AggregateOperator + Clone
 {
-    stack: VecDeque<Value>,
-    agg: Value,
-    op: PhantomData<BinOp>,
+    stack: VecDeque<BinOp::Partial>,
```

Comment:
Can we rename this from stack to queue? That would be more intuitive, since we use it as a FIFO, not a LIFO.

Reply:
Ditto.